server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3831,14 +3831,17 @@ public void afterRefresh(boolean didRefresh) {
      * @param listener the listener to invoke once the pending refresh location is visible. The listener will be called with
      *                 <code>true</code> if the listener was registered to wait for a refresh.
      */
-    public final void awaitShardSearchActive(Consumer<Boolean> listener) {
+    public final void makeShardSearchActive(Consumer<Boolean> listener) {
         markSearcherAccessed(); // move the shard into non-search idle
```

Comment:
Can we assert that we are not on a transport thread here (or that we are on a search or refresh thread)? Also, probably worth checking in more detail by checking callers (did not do so).

Reply:
Quickly looking this method is used by get, mget, explain and term vector transport actions too. So I guess we need to allow this method for these apis as well. Hopefully these api don't call this method from transport thread...

Reply:
Looks like `TransportOpenPointInTimeAction` calls this method from a transport thread.

Reply:
Unfortunately the `ensureShardSearchActive(...)` method (before `awaitShardSearchActive(...)`) is called in more places from transport_worker thread.

After adding the assertion we talked about, this assertion tripped in many places. For example in the `SearchService#executeQueryPhase(...)` case:

```
Caused by: java.lang.AssertionError: elasticsearch[node-0][transport_worker][T#12] not in [search, get] nor a test thread
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.threadpool.ThreadPool.assertCurrentThreadPool(ThreadPool.java:936)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.ensureShardSearchActive(IndexShard.java:3839)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$rewriteAndFetchShardRequest$15(SearchService.java:1632)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:158)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:112)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:77)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.search.SearchService.rewriteAndFetchShardRequest(SearchService.java:1638)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:477)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$8(SearchTransportService.java:490)
»  	at org.elasticsearch.security@8.9.0-SNAPSHOT/org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.doRun(SecurityServerTransportInterceptor.java:570)
»  	at org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
```

(note that in the `SearchService#executeQueryPhase(...)` case in a latter stage the operation continues on a search thread asynchronously, as part of executing the `ensureAfterSeqNoRefreshed(...)` method).

Maybe the refresh in the `ensureShardSearchActive(...)` method should be invoked from a refresh thread? Like this:

```
if (location == pendingRefreshLocation.get()) {
   threadPool.executor(ThreadPool.Names.REFRESH).execute(() -> {
      if (location == pendingRefreshLocation.get()) {
         getEngine().maybeRefresh("ensure-shard-search-active");
      }
   });
}
```

Reply:
Refresh now is always performed on a refresh thread: 803832e0db7225b5c7c27b5da26beabcd8305773